File: src/ios/FantasyGMAssistant/NativeModules/Media/MediaProcessor.m
//
// MediaProcessor.m
// FantasyGMAssistant
//
// Enhanced media processing bridge implementation for iOS
// Version: 1.0.0
//

#import <Foundation/Foundation.h> // iOS 14.0+
#import <React/RCTBridgeModule.h> // React Native 0.72+
#import "MediaProcessorBridge.h"
#import <FantasyGMAssistant-Swift.h> // Generated Swift header

// Error domain constant
NSString * const kMediaProcessorErrorDomain = @"com.fantasygm.mediaprocessor";

@interface MediaProcessorBridge ()

@property (nonatomic, strong) MediaProcessor *mediaProcessor;
@property (nonatomic, strong) NSMutableDictionary *activeOperations;
@property (nonatomic) dispatch_queue_t processingQueue;

@end

@implementation MediaProcessorBridge

RCT_EXPORT_MODULE(MediaProcessorBridge)

- (instancetype)init {
    if (self = [super init]) {
        // Initialize Swift MediaProcessor with default configuration
        self.mediaProcessor = [[MediaProcessor alloc] initWithConfig:nil];
        
        // Create serial queue for media processing
        self.processingQueue = dispatch_queue_create("com.fantasygm.mediaprocessor", 
                                                   DISPATCH_QUEUE_SERIAL);
        dispatch_set_target_queue(self.processingQueue, 
                                dispatch_get_global_queue(QOS_CLASS_USER_INITIATED, 0));
        
        // Initialize active operations tracking
        self.activeOperations = [NSMutableDictionary dictionary];
        
        // Register for memory warnings
        [[NSNotificationCenter defaultCenter] addObserver:self
                                               selector:@selector(handleMemoryWarning)
                                                   name:UIApplicationDidReceiveMemoryWarningNotification
                                                 object:nil];
    }
    return self;
}

+ (BOOL)requiresMainQueueSetup {
    return NO;
}

- (dispatch_queue_t)methodQueue {
    return self.processingQueue;
}

RCT_EXPORT_METHOD(generateTradeAnalysisVideo:(NSDictionary *)tradeDetails
                  withProgress:(RCTResponseSenderBlock)progressCallback
                  resolver:(RCTPromiseResolveBlock)resolve
                  rejecter:(RCTPromiseRejectBlock)reject) {
    
    NSString *operationId = [[NSUUID UUID] UUIDString];
    self.activeOperations[operationId] = @YES;
    
    // Validate input parameters
    if (!tradeDetails || ![tradeDetails isKindOfClass:[NSDictionary class]]) {
        reject(@"invalid_input", @"Invalid trade details provided", nil);
        return;
    }
    
    // Create progress handler
    void (^progressHandler)(double) = ^(double progress) {
        if (progressCallback) {
            progressCallback(@[@(progress)]);
        }
    };
    
    // Execute on processing queue
    dispatch_async(self.processingQueue, ^{
        [self.mediaProcessor generateTradeAnalysisVideoWithTradeDetails:tradeDetails 
                                                      progressHandler:progressHandler
                                                       completionHandler:^(NSURL * _Nullable videoURL, NSError * _Nullable error) {
            // Remove operation tracking
            [self.activeOperations removeObjectForKey:operationId];
            
            if (error) {
                reject(@"processing_failed",
                      error.localizedDescription,
                      error);
                return;
            }
            
            if (!videoURL) {
                reject(@"no_output",
                      @"Failed to generate video output",
                      nil);
                return;
            }
            
            resolve(videoURL.absoluteString);
        }];
    });
}

RCT_EXPORT_METHOD(generateVoiceOver:(NSString *)text
                  voiceSettings:(NSDictionary *)settings
                  resolver:(RCTPromiseResolveBlock)resolve
                  rejecter:(RCTPromiseRejectBlock)reject) {
    
    // Validate input text
    if (!text || [text length] == 0) {
        reject(@"invalid_input", @"Text content is required", nil);
        return;
    }
    
    // Configure voice quality
    VoiceQuality quality = VoiceQualityStandard;
    NSString *qualitySetting = settings[@"quality"];
    if ([qualitySetting isEqualToString:@"premium"]) {
        quality = VoiceQualityPremium;
    } else if ([qualitySetting isEqualToString:@"high"]) {
        quality = VoiceQualityHigh;
    }
    
    // Execute voice generation
    dispatch_async(self.processingQueue, ^{
        [self.mediaProcessor generateVoiceOverWithText:text
                                            quality:quality
                                  completionHandler:^(NSURL * _Nullable audioURL, NSError * _Nullable error) {
            if (error) {
                reject(@"voice_generation_failed",
                      error.localizedDescription,
                      error);
                return;
            }
            
            resolve(audioURL.absoluteString);
        }];
    });
}

RCT_EXPORT_METHOD(processMediaFile:(NSString *)fileURLString
                  options:(NSDictionary *)options
                  withProgress:(RCTResponseSenderBlock)progressCallback
                  resolver:(RCTPromiseResolveBlock)resolve
                  rejecter:(RCTPromiseRejectBlock)reject) {
    
    NSURL *fileURL = [NSURL URLWithString:fileURLString];
    if (!fileURL) {
        reject(@"invalid_url", @"Invalid file URL provided", nil);
        return;
    }
    
    // Create processing options
    MediaProcessingOptions *processingOptions = [[MediaProcessingOptions alloc] initWithDictionary:options];
    
    // Create progress handler
    void (^progressHandler)(double) = ^(double progress) {
        if (progressCallback) {
            progressCallback(@[@(progress)]);
        }
    };
    
    // Execute processing
    dispatch_async(self.processingQueue, ^{
        [self.mediaProcessor processMediaFileWithFileURL:fileURL
                                              options:processingOptions
                                      progressHandler:progressHandler
                                   completionHandler:^(NSURL * _Nullable outputURL, NSError * _Nullable error) {
            if (error) {
                reject(@"processing_failed",
                      error.localizedDescription,
                      error);
                return;
            }
            
            resolve(outputURL.absoluteString);
        }];
    });
}

RCT_EXPORT_METHOD(cancelOperation:(NSString *)operationId
                  resolver:(RCTPromiseResolveBlock)resolve
                  rejecter:(RCTPromiseRejectBlock)reject) {
    
    if ([self.activeOperations objectForKey:operationId]) {
        [self.mediaProcessor cancelCurrentTask];
        [self.activeOperations removeObjectForKey:operationId];
        resolve(@YES);
    } else {
        resolve(@NO);
    }
}

- (void)handleMemoryWarning {
    // Cancel non-critical operations
    [self.mediaProcessor cancelCurrentTask];
    
    // Clear active operations
    [self.activeOperations removeAllObjects];
    
    // Clean up temporary files
    [self.mediaProcessor cleanupTemporaryFiles];
}

- (void)dealloc {
    [[NSNotificationCenter defaultCenter] removeObserver:self];
}

@end

File: src/ios/FantasyGMAssistant/NativeModules/Media/MediaProcessor.swift
//
// MediaProcessor.swift
// FantasyGMAssistant
//
// Enhanced media processing implementation for trade analysis videos and voice synthesis
// Version: 1.0.0
//

import Foundation // iOS 14.0+
import AVFoundation // iOS 14.0+
import FFmpegKit // v5.1

// MARK: - Error Types
public enum MediaError: Error {
    case invalidInput
    case processingFailed
    case apiError(String)
    case resourceUnavailable
    case memoryWarning
    case cacheError
    case networkError
    case optimizationFailed
}

// MARK: - Type Definitions
public typealias ProgressHandler = (Double) -> Void

public enum VoiceQuality {
    case standard
    case high
    case premium
}

public struct MediaProcessingOptions {
    let quality: String
    let format: String
    let optimization: String
    let maxDuration: TimeInterval
    let maxFileSize: Int64
}

// MARK: - Resource Monitor
private class ResourceMonitor {
    static let shared = ResourceMonitor()
    private let memoryThreshold: Float = 0.8 // 80% memory usage threshold
    
    func checkResources() -> Bool {
        let memoryUsage = getMemoryUsage()
        return memoryUsage < memoryThreshold
    }
    
    private func getMemoryUsage() -> Float {
        var info = mach_task_basic_info()
        var count = mach_msg_type_number_t(MemoryLayout<mach_task_basic_info>.size)/4
        
        let kerr: kern_return_t = withUnsafeMutablePointer(to: &info) {
            $0.withMemoryRebound(to: integer_t.self, capacity: 1) {
                task_info(mach_task_self_, task_flavor_t(MACH_TASK_BASIC_INFO), $0, &count)
            }
        }
        
        return kerr == KERN_SUCCESS ? Float(info.resident_size) / Float(ProcessInfo.processInfo.physicalMemory) : 0.0
    }
}

// MARK: - Retry Manager
private class RetryManager {
    private let maxAttempts: Int = MAX_RETRY_ATTEMPTS
    
    func execute<T>(_ operation: @escaping () async throws -> T) async throws -> T {
        var lastError: Error?
        
        for attempt in 1...maxAttempts {
            do {
                return try await operation()
            } catch {
                lastError = error
                if attempt < maxAttempts {
                    try await Task.sleep(nanoseconds: UInt64(pow(2.0, Double(attempt)) * 1_000_000_000))
                }
            }
        }
        
        throw lastError ?? MediaError.processingFailed
    }
}

// MARK: - MediaProcessor Implementation
@available(iOS 14.0, *)
@objc public class MediaProcessor: NSObject {
    private let ffmpegKit: FFmpegKit
    private let cache: URLCache
    private let retryManager: RetryManager
    private let resourceMonitor: ResourceMonitor
    
    public init(config: MediaConfig? = nil) {
        self.ffmpegKit = FFmpegKit()
        self.cache = URLCache(memoryCapacity: MEDIA_CACHE_SIZE,
                            diskCapacity: MEDIA_CACHE_SIZE * 2,
                            directory: FileManager.default.temporaryDirectory.appendingPathComponent("MediaCache"))
        self.retryManager = RetryManager()
        self.resourceMonitor = ResourceMonitor.shared
        super.init()
        
        NotificationCenter.default.addObserver(self,
                                             selector: #selector(handleMemoryWarning),
                                             name: UIApplication.didReceiveMemoryWarningNotification,
                                             object: nil)
    }
    
    // MARK: - Public Methods
    public func generateTradeAnalysisVideo(tradeDetails: [String: Any],
                                         progressHandler: ProgressHandler? = nil) async -> Result<URL, MediaError> {
        guard resourceMonitor.checkResources() else {
            return .failure(.resourceUnavailable)
        }
        
        do {
            // Validate input
            guard let playersInvolved = tradeDetails["players"] as? [[String: Any]],
                  !playersInvolved.isEmpty else {
                return .failure(.invalidInput)
            }
            
            progressHandler?(0.1)
            
            // Generate script
            let script = try await generateVideoScript(tradeDetails)
            progressHandler?(0.2)
            
            // Generate voiceover
            let voiceoverResult = await generateVoiceOver(text: script, quality: .premium)
            guard case .success(let audioURL) = voiceoverResult else {
                return .failure(.processingFailed)
            }
            progressHandler?(0.4)
            
            // Generate video content
            let videoURL = try await generateVideoContent(tradeDetails, audioURL)
            progressHandler?(0.7)
            
            // Process final video
            let options = MediaProcessingOptions(quality: "high",
                                               format: "mp4",
                                               optimization: "quality",
                                               maxDuration: 180,
                                               maxFileSize: 50_000_000)
            
            let result = await processMediaFile(fileURL: videoURL,
                                              options: options,
                                              progressHandler: { progress in
                progressHandler?(0.7 + (progress * 0.3))
            })
            
            cleanupTemporaryFiles()
            return result
            
        } catch {
            cleanupTemporaryFiles()
            return .failure(.processingFailed)
        }
    }
    
    public func generateVoiceOver(text: String,
                                quality: VoiceQuality) async -> Result<URL, MediaError> {
        guard !text.isEmpty, text.count <= 5000 else {
            return .failure(.invalidInput)
        }
        
        return await retryManager.execute {
            let endpoint = Constants.APIEndpoints.media.voiceover
            let parameters: [String: Any] = [
                "text": text,
                "quality": quality.rawValue,
                "api_key": ELEVEN_LABS_API_KEY
            ]
            
            // API call implementation
            let (data, response) = try await URLSession.shared.upload(
                for: URLRequest(url: URL(string: endpoint)!),
                from: try JSONSerialization.data(withJSONObject: parameters)
            )
            
            guard let httpResponse = response as? HTTPURLResponse,
                  httpResponse.statusCode == 200 else {
                throw MediaError.apiError("Voice generation failed")
            }
            
            let outputURL = FileManager.default.temporaryDirectory
                .appendingPathComponent(UUID().uuidString)
                .appendingPathExtension("mp3")
            
            try data.write(to: outputURL)
            return .success(outputURL)
        } ?? .failure(.processingFailed)
    }
    
    public func processMediaFile(fileURL: URL,
                               options: MediaProcessingOptions,
                               progressHandler: ProgressHandler? = nil) async -> Result<URL, MediaError> {
        guard FileManager.default.fileExists(atPath: fileURL.path) else {
            return .failure(.invalidInput)
        }
        
        do {
            let outputURL = FileManager.default.temporaryDirectory
                .appendingPathComponent(UUID().uuidString)
                .appendingPathExtension(options.format)
            
            let command = generateFFmpegCommand(input: fileURL,
                                              output: outputURL,
                                              options: options)
            
            let session = FFmpegKit.execute(command)
            guard session?.getReturnCode().isValueSuccess() ?? false else {
                throw MediaError.processingFailed
            }
            
            return .success(outputURL)
        } catch {
            return .failure(.processingFailed)
        }
    }
    
    // MARK: - Private Methods
    private func generateVideoScript(_ tradeDetails: [String: Any]) async throws -> String {
        // Implementation of script generation logic
        return ""
    }
    
    private func generateVideoContent(_ tradeDetails: [String: Any], _ audioURL: URL) async throws -> URL {
        // Implementation of video content generation logic
        return URL(fileURLWithPath: "")
    }
    
    private func generateFFmpegCommand(input: URL, output: URL, options: MediaProcessingOptions) -> String {
        var command = "-i \(input.path) "
        
        switch options.quality {
        case "high":
            command += "-c:v libx264 -preset slow -crf 22 "
        case "medium":
            command += "-c:v libx264 -preset medium -crf 23 "
        default:
            command += "-c:v libx264 -preset fast -crf 24 "
        }
        
        command += "-c:a aac -b:a 128k "
        command += "\(output.path)"
        
        return command
    }
    
    private func cleanupTemporaryFiles() {
        let fileManager = FileManager.default
        let tmpDirectory = fileManager.temporaryDirectory
        
        do {
            let tmpContents = try fileManager.contentsOfDirectory(at: tmpDirectory,
                                                                includingPropertiesForKeys: nil,
                                                                options: [])
            for file in tmpContents where file.pathExtension == "mp4" || file.pathExtension == "mp3" {
                try? fileManager.removeItem(at: file)
            }
        } catch {
            print("Cleanup error: \(error)")
        }
    }
    
    @objc private func handleMemoryWarning() {
        cache.removeAllCachedResponses()
        cleanupTemporaryFiles()
    }
    
    deinit {
        NotificationCenter.default.removeObserver(self)
    }
}